File: 02_selection_sort/ruby/01_selection_sort.rb
# Finds the smallest value in an array
def find_smallest(arr)
  # Stores the smallest value
  smallest = arr[0]
  # Stores the index of the smallest value
  smallest_index = 0
  # (1...n) is the same as (1..(n-1))
  (1...(arr.length)).each do |i|
    if arr[i] < smallest
      smallest = arr[i]
      smallest_index = i
    end
  end
  smallest_index
end

# Sort array
def selection_sort(arr)
  new_arr = []
  arr.length.times do
    # Finds the smallest element in the array and adds it to the new array
    smallest = find_smallest(arr)
    new_arr.push(arr.delete_at(smallest))
  end
  new_arr
end

# 'p obj' is the same as 'puts obj.inspect'
p selection_sort([5, 3, 6, 2, 10])

File: 02_selection_sort/swift/01_selection_sort.swift
import Foundation

// Finds the smallest value in an array
func findSmallestIndex <T: Comparable> (_ arr: [T]) -> Int {
    // Stores the smallest value
    var smallest = arr[0]
    // We don't need any calculation if the array lenght is 1
    if arr.count == 1 {
        return 0
    }
    // Stores the index of the smallest value
    var smallestIndex = 0
    for i in 1...arr.count-1 {
        if arr[i] < smallest {
            smallest = arr[i]
            smallestIndex = i
        }
    }
    return smallestIndex
}

// Sort array
func selectionSort <T: Comparable> (arr: [T]) -> [T] {
    var newArr: [T] = []
    // We have to make mutableArray reference copy of original array, because Swift 3 doesn't allow to get var parameter
    var mutableArr = arr
    for _ in 0...mutableArr.count-1 {
       //Finds the smallest element in the array and adds it to the new array
        let smallestIndex = findSmallestIndex(mutableArr)
        newArr.append(mutableArr[smallestIndex])
        mutableArr.remove(at: smallestIndex)
    }
    return newArr
}

print(selectionSort(arr: [5, 3, 6, 2, 10])) // => [2, 3, 5, 6, 10]


